File: backends/micronaut-java/src/main/java/dev/stackverse/backend/JwtVerifier.java
package dev.stackverse.backend;

import com.nimbusds.jose.JWSAlgorithm;
import com.nimbusds.jose.crypto.RSASSAVerifier;
import com.nimbusds.jose.jwk.JWKSet;
import com.nimbusds.jose.jwk.RSAKey;
import com.nimbusds.jwt.JWTClaimsSet;
import com.nimbusds.jwt.SignedJWT;
import io.micronaut.context.annotation.Value;
import jakarta.inject.Singleton;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;
import java.time.Instant;
import java.util.ArrayList;
import java.util.Date;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import tools.jackson.databind.JsonNode;
import tools.jackson.databind.ObjectMapper;

@Singleton
final class JwtVerifier {
    private static final Logger LOG = LoggerFactory.getLogger(JwtVerifier.class);
    private static final Duration REFRESH_COOLDOWN = Duration.ofSeconds(30);

    private final String issuer;
    private final String configuredJwksUri;
    private final String audience;
    private final ObjectMapper mapper;
    private final HttpClient client = HttpClient.newBuilder().connectTimeout(Duration.ofSeconds(10)).build();
    private final Object refreshLock = new Object();
    private volatile Map<String, RSAKey> keys = Map.of();
    private volatile Instant lastRefresh = Instant.EPOCH;
    private volatile String resolvedJwksUri;

    JwtVerifier(
            @Value("${stackverse.oidc.issuer-uri}") String issuer,
            @Value("${stackverse.oidc.jwks-uri:}") String configuredJwksUri,
            @Value("${stackverse.oidc.audience}") String audience,
            ObjectMapper mapper
    ) {
        this.issuer = issuer;
        this.configuredJwksUri = configuredJwksUri;
        this.audience = audience;
        this.mapper = mapper;
    }

    Identity verify(String raw) {
        try {
            SignedJWT jwt = SignedJWT.parse(raw);
            if (!List.of(JWSAlgorithm.RS256, JWSAlgorithm.RS384, JWSAlgorithm.RS512).contains(jwt.getHeader().getAlgorithm())) {
                throw new IllegalArgumentException("unsupported algorithm");
            }
            RSAKey key = key(jwt.getHeader().getKeyID());
            if (!jwt.verify(new RSASSAVerifier(key.toRSAPublicKey()))) {
                throw new IllegalArgumentException("bad signature");
            }
            JWTClaimsSet claims = jwt.getJWTClaimsSet();
            Date expires = claims.getExpirationTime();
            if (expires == null || expires.toInstant().isBefore(Instant.now())) {
                throw new IllegalArgumentException("token expired");
            }
            if (!issuer.equals(claims.getIssuer())) {
                throw new IllegalArgumentException("bad issuer");
            }
            if (claims.getAudience() == null || !claims.getAudience().contains(audience)) {
                throw new IllegalArgumentException("bad audience");
            }
            String username = claims.getStringClaim("preferred_username");
            if (username == null || username.isBlank()) {
                throw new IllegalArgumentException("missing preferred_username");
            }
            return new Identity(username, claims.getStringClaim("name"), claims.getStringClaim("email"), roles(claims));
        } catch (Exception ex) {
            throw new IllegalArgumentException("invalid token", ex);
        }
    }

    @SuppressWarnings("unchecked")
    private List<String> roles(JWTClaimsSet claims) {
        try {
            Object realm = claims.getClaim("realm_access");
            if (!(realm instanceof Map<?, ?> map)) {
                return List.of();
            }
            Object roles = map.get("roles");
            if (!(roles instanceof List<?> list)) {
                return List.of();
            }
            List<String> result = new ArrayList<>();
            for (Object role : list) {
                if (role instanceof String value) {
                    if ("moderator".equals(value) || "admin".equals(value)) {
                        result.add(value);
                    }
                }
            }
            return result;
        } catch (RuntimeException ex) {
            return List.of();
        }
    }

    private RSAKey key(String kid) throws Exception {
        if (kid == null || kid.isBlank()) {
            throw new IllegalArgumentException("missing signing key id");
        }
        RSAKey existing = keys.get(kid);
        if (existing != null) {
            return existing;
        }
        if (Duration.between(lastRefresh, Instant.now()).compareTo(REFRESH_COOLDOWN) < 0) {
            throw new IllegalArgumentException("unknown signing key");
        }
        Map<String, RSAKey> refreshed = fetchKeys();
        synchronized (refreshLock) {
            keys = Map.copyOf(refreshed);
            lastRefresh = Instant.now();
        }
        RSAKey key = keys.get(kid);
        if (key == null) {
            throw new IllegalArgumentException("unknown signing key");
        }
        return key;
    }

    private Map<String, RSAKey> fetchKeys() throws Exception {
        String uri = jwksUri();
        long started = System.nanoTime();
        try {
            HttpRequest request = HttpRequest.newBuilder()
                    .uri(URI.create(uri))
                    .timeout(Duration.ofSeconds(10))
                    .GET()
                    .build();
            HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());
            if (response.statusCode() != 200) {
                throw new IllegalStateException("JWKS endpoint returned " + response.statusCode());
            }
            JWKSet set = JWKSet.parse(response.body());
            Map<String, RSAKey> refreshed = new HashMap<>();
            for (RSAKey key : set.getKeys().stream().filter(jwk -> jwk instanceof RSAKey).map(jwk -> (RSAKey) jwk).toList()) {
                if (key.getKeyID() != null && !key.getKeyID().isBlank()) {
                    refreshed.put(key.getKeyID(), key);
                }
            }
            if (refreshed.isEmpty()) {
                throw new IllegalStateException("JWKS contains no RSA signing keys");
            }
            return refreshed;
        } catch (Exception ex) {
            EventLog.warn(LOG, "dependency_call_failed", "failure", "Fetching JWKS failed",
                    Map.of("dependency", "keycloak", "duration_ms", (System.nanoTime() - started) / 1_000_000,
                            "error_code", "jwks_fetch_failed"));
            throw ex;
        }
    }

    private String jwksUri() throws Exception {
        if (resolvedJwksUri != null) {
            return resolvedJwksUri;
        }
        if (configuredJwksUri != null && !configuredJwksUri.isBlank()) {
            resolvedJwksUri = configuredJwksUri;
            return resolvedJwksUri;
        }
        long started = System.nanoTime();
        try {
            HttpRequest request = HttpRequest.newBuilder()
                    .uri(URI.create(issuer + "/.well-known/openid-configuration"))
                    .timeout(Duration.ofSeconds(10))
                    .GET()
                    .build();
            HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());
            if (response.statusCode() != 200) {
                throw new IllegalStateException("OIDC discovery returned " + response.statusCode());
            }
            JsonNode root = mapper.readTree(response.body());
            String discovered = root.path("jwks_uri").asText("");
            if (discovered.isBlank()) {
                throw new IllegalStateException("OIDC discovery response has no jwks_uri");
            }
            resolvedJwksUri = discovered;
            return discovered;
        } catch (Exception ex) {
            EventLog.warn(LOG, "dependency_call_failed", "failure", "OIDC discovery failed",
                    Map.of("dependency", "keycloak", "duration_ms", (System.nanoTime() - started) / 1_000_000,
                            "error_code", "oidc_discovery_failed"));
            throw ex;
        }
    }
}

File: backends/php-laravel/app/Auth/JwtVerifier.php
<?php

namespace App\Auth;

use App\Support\Logger;
use Firebase\JWT\JWK;
use Firebase\JWT\JWT;
use Illuminate\Support\Facades\Cache;
use Illuminate\Support\Facades\Http;
use RuntimeException;

class JwtVerifier
{
    private const LEEWAY_SECONDS = 30;

    public function verify(string $token): Caller
    {
        JWT::$leeway = self::LEEWAY_SECONDS;
        $decoded = JWT::decode($token, JWK::parseKeySet($this->jwks(), 'RS256'));
        $payload = json_decode(json_encode($decoded, JSON_THROW_ON_ERROR), true, flags: JSON_THROW_ON_ERROR);

        $this->validateClaims($payload);
        $username = $payload['preferred_username'] ?? null;
        if (! is_string($username) || $username === '') {
            throw new RuntimeException('missing preferred_username');
        }

        $roles = $payload['realm_access']['roles'] ?? [];
        if (! is_array($roles)) {
            $roles = [];
        }

        return new Caller(
            $username,
            array_values(array_filter($roles, static fn (mixed $role): bool => is_string($role))),
            isset($payload['name']) && is_string($payload['name']) ? $payload['name'] : null,
            isset($payload['email']) && is_string($payload['email']) ? $payload['email'] : null,
        );
    }

    private function validateClaims(array $payload): void
    {
        $issuer = config('stackverse.oidc.issuer_uri');
        if (($payload['iss'] ?? null) !== $issuer) {
            throw new RuntimeException('issuer mismatch');
        }

        $audience = config('stackverse.oidc.audience');
        $aud = $payload['aud'] ?? null;
        $audiences = is_array($aud) ? $aud : [$aud];
        if (! in_array($audience, $audiences, true)) {
            throw new RuntimeException('audience mismatch');
        }

        $now = time();
        if (! isset($payload['exp']) || (int) $payload['exp'] < $now - self::LEEWAY_SECONDS) {
            throw new RuntimeException('token expired');
        }
        if (isset($payload['nbf']) && (int) $payload['nbf'] > $now + self::LEEWAY_SECONDS) {
            throw new RuntimeException('token not yet valid');
        }
    }

    private function jwks(): array
    {
        return Cache::remember('stackverse.oidc.jwks', now()->addMinutes(5), fn (): array => $this->fetchJwks());
    }

    private function fetchJwks(): array
    {
        $jwksUri = config('stackverse.oidc.jwks_uri');
        if (! is_string($jwksUri) || $jwksUri === '') {
            $discovery = rtrim((string) config('stackverse.oidc.issuer_uri'), '/').'/.well-known/openid-configuration';
            $started = microtime(true);
            try {
                $response = Http::timeout(5)->get($discovery);
                if (! $response->successful()) {
                    throw new RuntimeException('discovery answered '.$response->status());
                }
                $jwksUri = $response->json('jwks_uri');
            } catch (\Throwable $error) {
                Logger::event('error', 'dependency_call_failed', 'failure', 'OIDC discovery failed', [
                    'dependency' => 'keycloak',
                    'duration_ms' => (int) round((microtime(true) - $started) * 1000),
                    'error_code' => 'oidc_discovery_failed',
                ]);

                throw $error;
            }
        }

        $started = microtime(true);
        try {
            $response = Http::timeout(5)->get($jwksUri);
            if (! $response->successful()) {
                throw new RuntimeException('jwks answered '.$response->status());
            }

            $jwks = $response->json();
            if (! is_array($jwks)) {
                throw new RuntimeException('jwks response was not an object');
            }

            return $jwks;
        } catch (\Throwable $error) {
            Logger::event('error', 'dependency_call_failed', 'failure', 'JWKS fetch failed', [
                'dependency' => 'keycloak',
                'duration_ms' => (int) round((microtime(true) - $started) * 1000),
                'error_code' => 'jwks_fetch_failed',
            ]);

            throw $error;
        }
    }
}
